pandas/plotting/_matplotlib/groupby.py

```diff
@@ -0,0 +1,136 @@
+    If `by` is assigned:
+
+    >>> import numpy as np
+    >>> tuples = [('h1', 'a'), ('h1', 'b'), ('h2', 'a'), ('h2', 'b')]
+    >>> mi = MultiIndex.from_tuples(tuples)
+    >>> value = [[1, 3, np.nan, np.nan],
+    ...          [3, 4, np.nan, np.nan], [np.nan, np.nan, 5, 6]]
+    >>> data = DataFrame(value, columns=mi)
+    >>> create_iter_data_given_by(data, by=["col1"])
+    {'h1': DataFrame({'a': [1, 3, np.nan], 'b': [3, 4, np.nan]}),
+     'h2': DataFrame({'a': [np.nan, np.nan, 5], 'b': [np.nan, np.nan, 6]})}
+    """
+    if kind == "hist":
+        level = 0
+    elif kind == "box":
+        level = 1
+    else:
+        raise ValueError("This function is only used for hist and box plot")
+
+    iter_data: Union[DataFrame, Dict[str, FrameOrSeriesUnion]]
+    if not by:
+        iter_data = data
+    else:
+        # Select sub-columns based on the value of first level of MI
+        assert isinstance(data.columns, MultiIndex)
+        cols = data.columns.levels[level]
+        iter_data = {
+            col: data.loc[:, data.columns.get_level_values(level) == col]
+            for col in cols
+        }
+    return iter_data
+
+
+def reconstruct_data_with_by(
+    data: DataFrame, by: IndexLabel, cols: IndexLabel
+) -> DataFrame:
+    """
+    Internal function to group data, and reassign multiindex column names onto the
+    result in order to let grouped data be used in _compute_plot_data method.
+
+    Parameters
+    ----------
+    data: Original DataFrame to plot
+    by: grouped `by` parameter selected by users
+    cols: columns of data set (excluding columns used in `by`)
+
+    Returns
+    -------
+    Output is the reconstructed DataFrame with MultiIndex columns. The first level
+    of MI is unique values of groups, and second level of MI is the columns
+    selected by users.
+
+    Examples
+    --------
+    >>> d = {'h': ['h1', 'h1', 'h2'], 'a': [1, 3, 5], 'b': [3, 4, 6]}
+    >>> df = DataFrame(d)
+    >>> reconstruct_data_with_by(df, by='h', cols=['a', 'b'])
+       h1      h2
+       a   b   a   b
+    0  1   3   NaN NaN
+    1  3   4   NaN NaN
+    2  NaN NaN 5   6
+    """
+    grouped = data.groupby(by)
+
+    data_list = []
+    for key, group in grouped:
+        columns = MultiIndex.from_product([[key], cols])
+        sub_group = group[cols]
+        sub_group.columns = columns
+        data_list.append(sub_group)
+
+    data = concat(data_list, axis=1)
+    return data
+
+
+def reformat_hist_y_given_by(
+    y: Union[Series, np.ndarray], by: Optional[IndexLabel] = None
+) -> Union[Series, np.ndarray]:
+    """Internal function to reformat y given `by` is applied or not for hist plot.
+
+    If by is None, input y is 1-d with NaN removed; and if by is not None, groupby
+    will take place and input y is multi-dimensional array.
+    """
+    if by is not None and len(y.shape) > 1:
+        y = np.array([remove_na_arraylike(col) for col in y.T]).T
+    else:
+        y = remove_na_arraylike(y)
+    return y
```

Comment:
Small thing, and maybe just personal preference, but I'd write this as:
```python
if by is not None and len(y.shape) > 1:
    return np.array([remove_na_arraylike(col) for col in y.T]).T
return remove_na_arraylike(y)
```
I find it a bit distracting being it so verbose for not much value.

Reply:
Sure!